### DISCUSSION TITLE
I2S input from second ESP32 (Karadio32) to send audio to a BT Speaker

### DISCUSSION THREAD

**Discussion by JocB1** [pschatzmann/ESP32-A2DP#281] [Q&A] [answered]
Hi, 

I have built several Karadio32 Internet radios which use the ESP32. Partly also with I2S DAC or I2S Amp (MAX98xx). 

Now I need a way to stream the I2S output signal of the ESP32 (Karadio32) via Bluetooth to my BT speaker at the Tereasse. I think this project would fit if I donate a second ESP32 that only takes care of the BT audio transmission. 

Is this possible and can someone give me some guidance on how to configure this?

Best regards, 

Jochen

**Comment by JocB1** [comment]
I mean a example for BT source with I2S input signals, not BT sink with I2S output.

**Comment by pschatzmann** [comment, marked as answer]
Here is an example which is part of the arduino-audio-tools: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-basic-api/base-i2s-a2dp

However I think the challenge might be to get a stable I2S between 2 ESP32. I have never tried this myself, but if you look in the project there are some discussions and issues of failed attempts. My gut feeling is, that it might be easier to use a simple Serial connection and maybe transmit the encoded data.

**Comment by JocB1** [comment]
I tried the example sending a sine wave to the BT speaker, that worked fine. 

The I2S in the BT example is designed to read an i2S microphone, i.e. the master (BT ESP) sends the clock to the slave (microphone). 
In the Karadio32 project it is so that an I2S DAC is controlled, i.e. the clock is generated by the Karadio32 ESP and must be sent to an input of the BT ESP. The same applies to word select. 
Is there a way to set this up?

**Comment by pschatzmann** [comment]
Just use 
  cfg.is_master = false;
This way the pin_bck and pin_ws will be used as input. 
Note however that the ESP32 does not support the pin_mck as input!

https://github.com/pschatzmann/arduino-audio-tools/issues/63

**Comment by JocB1** [comment]
sorry for asking but where exactly can i find the setting "cfg.is_master = xx;"?

 
 

Gesendet: Donnerstag, 11. August 2022 um 04:07 Uhr
Von: "Phil Schatzmann" ***@***.***>
An: "pschatzmann/ESP32-A2DP" ***@***.***>
Cc: "JocB1" ***@***.***>, "Author" ***@***.***>
Betreff: Re: [pschatzmann/ESP32-A2DP] I2S input from second ESP32 (Karadio32) to send audio to a BT Speaker (Discussion #281)


 

Just use
cfg.is_master = false;
This way the pin_bck and pin_ws will be used as input.
Note however that the ESP32 does not support the pin_mck as input!

**Comment by pschatzmann** [comment]
Well, to the example that I have sent you 5 days ago. The default setting is cfg.is_master = true; so it has been omitted like many other settings as well.

**Comment by JocB1** [comment]
OK,
I was in the wrong part...
Now I used your example and changed the settings:
...
  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT                   ---- ????
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate        = 48000;           // found out with logic analyzer
  cfg.is_master          = false;     
  i2s.begin(cfg);

  // start the bluetooth
  Serial.println("starting A2DP...");

  a2dp_source.start("BT-WE", get_sound_data);                  // This speaker connects

  //a2dp_source.start("JBL Flip 4", get_sound_data);          // This speaker connects not - but this one is to be used later!
}

...

For finding out the sample rate I coud do some meassurements with a logic analyzer connected to the Karadio32 output (I2S setting, not MERUS setting:


![I2S_Karadio32] (https://user-images.githubusercontent.com/1819346/186661755-a9fbcd8b-1ad9-43d2-abdb-bb9c5e0e2fd4.png)

See the meassure markers on the right side: 
Clock frequency: 1.56 MHz 
Sample Rate:  47.98 kHz 


The ESP32 connects to the BT speaker and the speaker signals this with a corresponding sound.

However, the sound from the Karadio32 is not transmitted. In the serial monitor you can see that the ESP restarts again and again (Guru Meditation).

Do you have any idea what else i can try?

**Comment by pschatzmann** [comment]
I suggest to take one step back: Did you manage to receive the I2S data and send it to CSVStream, so that you can check your I2S connection ? 

After the connections has started you can query the sample rate from A2DP. It should usually be 44100.

**Comment by JocB1** [comment]
I tested now the I2S receiving with this sketch: streams-i2s-serial_16bit.ino 

This is the output, but I´m not sure if this is correct: 

![Serial_Plot__I2S-to-16Bit](https://user-images.githubusercontent.com/1819346/187020018-efdb810e-511d-40fa-a0ff-c3becb645c6c.png)

....
>>> After the connections has started you can query the sample rate from A2DP. It should usually be 44100.

Do you mean with the "streams-generator-a2dp.ino" sketch to check only the A2DP connection? 
But how to query the Sample rate? 


The incomeing I2S sample Rate I had meassured with 48kHz (Logic Analyzer) so must the sample rate to the BT Speker be the same or could it be different for Incoming I2S and outgoing A2DP?

**Comment by pschatzmann** [comment]
This does not look like valid sound waves. Double check that the bits_per_sample are set up identical on both systems.
Double check in the documentation but I think it is:
a2dp_source.source().get_sample_rate()

**Comment by JocB1** [comment]
I think that looks better now
![Serial_Plot__I2S-to-16Bit__270822](https://user-images.githubusercontent.com/1819346/187021396-65b87917-3dba-43d8-a086-597f5ddf3b73.png)
..
    auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.i2s_format = I2S_STD_FORMAT;   // or try with I2S_LSB_FORMAT     -> should be "MSB first" because Karadio32 works with MAX98357 amp, and Datasheet says "MSB first" 
    cfg.bits_per_sample = 16;
    cfg.channels = 2;
    cfg.sample_rate = 48000;
    cfg.is_master = false;
     // this module nees a master clock if the ESP32 is master
    cfg.use_apll = true;    // false;  // try with yes
    //cfg.pin_mck = 3; 
..

**Comment by pschatzmann** [comment]
Hmm, this still looks very noisy to me: The volume also looks very low.
You could listen to this e.g. with a webserver example

**Comment by JocB1** [comment]
I guess the reboots of the ESP32 are related to this part: 
....
ConverterFillLeftAndRight<int16_t> bothChannels(LeftIsEmpty);
const size_t max_buffer_len = 150;
const int channels = 2;
const size_t max_buffer_bytes = max_buffer_len * sizeof(int16_t) * channels;
uint8_t buffer[max_buffer_bytes]={0};

// callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
int32_t get_sound_data(Frame* data, int32_t len) {
  size_t req_bytes = min(max_buffer_bytes, len*2*sizeof(int16_t));
  // the microphone provides data in int32_t -> we read it into the buffer of int32_t data so *2
  size_t result_bytes = i2s.readBytes((uint8_t*)buffer, req_bytes*2);
  // we have data only in 1 channel but we want to fill both
  return bothChannels.convert((uint8_t*)buffer, result_bytes);
}
...

This is relatad to a microphone with special characteristics, but we have a source that provides 2 channels. 

How would that have to be look if the two channels are separate?

**Reply by pschatzmann** [reply]
Well, just omit the converter

**Reply by JocB1** [reply]
Yes, that's clear as far as it goes, but how...? 


---
int32_t get_sound_data(Frame* data, int32_t len) 
{
  
  size_t req_bytes = min(max_buffer_bytes, len*2*sizeof(int16_t));
  // the microphone provides data in int32_t -> we read it into the buffer of int32_t data so *2
 
  // size_t result_bytes = i2s.readBytes((uint8_t*)buffer, req_bytes      );             // *2);
  // we have data only in 1 channel but we want to fill both  );  //
  
//  return   bothChannels.convert(    (uint8_t*)  buffer, result_bytes);                         // We don´t need the converter.... 

return   (  (int32_t)  i2s.readBytes((uint8_t*)buffer, req_bytes  *2) ) ;                      // ?? We need still both chanels -> *2 ?? Return value is int32_t

}
---

Am I on the right track?

**Comment by pschatzmann** [comment]
You just need to convert between frames and bytes
```
int32_t get_sound_data(Frame* data, int32_t frameCount) {
  int frameSize = 4; // 2 channels * 2 bytes 
  return i2s.readBytes((uint8_t*)data, frameCount *frameSize) / frameSize;
}
```

**Comment by JocB1** [comment]
Thanks, 
the ESP32 is now running without restarts - YES, one goal achieved! 

---
st:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1216
ho 0 tail 12 room 4
load:0x40078000,len:10944
load:0x40080400,len:6388
entry 0x400806b4
starting I2S...
[I] I2SConfig.h : 95 - rx/tx mode: RX
[I] I2SConfig.h : 96 - port_no: 0
[I] I2SConfig.h : 97 - is_master: Slave
[I] I2SConfig.h : 98 - sample rate: 48000
[I] I2SConfig.h : 99 - bits per sample: 16
[I] I2SConfig.h : 100 - number of channels: 2
[I] I2SConfig.h : 101 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 103 - auto_clear:0
[I] I2SConfig.h : 105 - use_apll: true
[I] I2SConfig.h : 110 - buffer_count:10
[I] I2SConfig.h : 111 - buffer_size:512
[I] I2SConfig.h : 115 - pin_bck: 14
[I] I2SConfig.h : 116 - pin_ws: 15
[I] I2SConfig.h : 117 - pin_data: 32
starting A2DP...
---

Puh, unfortunately there is still no sound from the speaker. 

>>> Double check in the documentation but I think it is:
>>> a2dp_source.source().get_sample_rate()

For BT Sink I´ve fond such a command in the docu but not for source (which we do have): 
https://pschatzmann.github.io/ESP32-A2DP/html/class_bluetooth_a2_d_p_source.html

One thought: From I2S the data comes with 48 kHz - can this be any problem? In the Sine -> A2DP example, I have successfully tested the BT sample rate at 48 kHz.

**Comment by pschatzmann** [comment]
You are right, I was mixing this up with the audio sink where we have this method: I am pretty sure that it needs to be 44100 and I used this in my examples, but I never tried with  48000.
And yes I think that it is critical that all values alligned across the whole chain: the I2S receiver can not slow down the sender when too much data is coming in too fast: the buffer will just overflow and data is lost

**Comment by JocB1** [comment]
Today I startet a new try: 

ESP32 #1: Example streams-generator-i2s (44.1 khz, 2 ch, 16bit - output clk 14 / ws 15 / data 22)

![I2S_Sinus_Generator___44_1_KHz__16Bit_2CH](https://user-images.githubusercontent.com/1819346/188283470-b6d74029-4792-44c0-a40f-646e28f21c00.png)
-> What comes out of the sinus generator as i2s signal looks plausible. 

ESP32 #2:  Base-I2S-A2DP (Input clock 14 / ws 15 / data 32): 

-----
/**
 * @file base-i2s-a2dp.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/base-i2s-a2dp/README.md
  * @copyright GPLv3
*/

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

/**
 * @brief We use a INMP441 I2S microphone as input and send the data to A2DP
 * Unfortunatly the data type from the microphone (int32_t)  does not match with the required data type by A2DP (int16_t),
 * so we need to convert.
 */ 

BluetoothA2DPSource a2dp_source;
I2SStream i2s;
ConverterFillLeftAndRight<int16_t> bothChannels(LeftIsEmpty);
const size_t max_buffer_len = 150;
const int channels = 2;
const size_t max_buffer_bytes = max_buffer_len * sizeof(int16_t) * channels;
uint8_t buffer[max_buffer_bytes]={0};

//     // callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
//      int32_t get_sound_data(Frame* data, int32_t len) {
//       size_t req_bytes = min(max_buffer_bytes, len*2*sizeof(int16_t));
//       // the microphone provides data in int32_t -> we read it into the buffer of int32_t data so *2
//        size_t result_bytes = i2s.readBytes((uint8_t*)buffer, req_bytes*2);
//        // we have data only in 1 channel but we want to fill both
//        return bothChannels.convert((uint8_t*)buffer, result_bytes);


// from Author for 2ch 16bit stereo (not a micro) 
int32_t get_sound_data(Frame* data, int32_t frameCount) {
  int frameSize = 4; // 2 channels * 2 bytes 
  return i2s.readBytes((uint8_t*)buffer, frameCount *frameSize) / frameSize;
}


// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate = 44100;
  cfg.is_master =  false;                   //                    true;
  i2s.begin(cfg);

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.start("BT-WE", get_sound_data);  
}

// Arduino loop - repeated processing 
void loop() {
}

------


The two ESP32´s are connected together with the same settings (2ch / 16bit / 44.1 kHz) 

-> Now we can assume that the sample rate is the same everywhere in the chain. 

The second ESP connects to the speaker, but there´s no sound.... 

PS: I use 2x ESP32 wrover - that should not be a problem, right?

**Comment by pschatzmann** [comment]
Wrover should work just fine and the setup looks right to me as well.
I think there is a logic in A2DP to switch off the sound if the signal is too noisy...

You can e.g. try to [send the audio to your browser](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-webserver/streams-i2s-webserver_wav/streams-i2s-webserver_wav.ino) and double check what this gives. Just remove the filter from the example...

**Comment by JocB1** [comment]
Hi, I cross-checked again and connected the I2S amplifier MAX98357 to the Sinus Generatur ESP. The sine wave comes out loud, clear and distinct. Noise can not be the problem: 

![ESP32_Sine_Generator___MAX98357](https://user-images.githubusercontent.com/1819346/188299833-5f72cf55-817c-4b8a-808e-79f674671348.jpg)


I guess that does not fit what with the ESP, which converts I2S to A2DP.

(Perplexed) 

----

I would still have the possibility at the Karadio32 the sound over spdif output. Is there a code for SPDIF to A2DP?

**Reply by Jeff-Jordan** [reply]
Hello JocB1,
did you finally succeed to connect two ESP32 thru I2S and pass the audio data to a bt-speaker?
If so, it would be nice to share the result.

I'm facing here the same problem....

Or maybe Phil could add it to his ESP32-A2DP samples?

**Comment by pschatzmann** [comment]
Oh, delete everything that is related to the buffer and converter which is not needed!
The callback should return the data:
return i2s.readBytes((uint8_t*) data, frameCount *frameSize) / frameSize;

With this correction it should work!
